### ISSUE TITLE
[flax.linen] Disable dropout in SelfAttention dynamically when using `setup`

### ISSUE THREAD

**Issue by patrickvonplaten** [google/flax#821]
### Problem you have encountered:

Dropout allows to be enabled/disabled within the forward call since `flax.linen.Dropout` accepts the boolean `deterministic` in its `__call__` function, see: https://flax.readthedocs.io/en/latest/_modules/flax/linen/stochastic.html#Dropout .

However the `flax.linen.SelfAttention` class does not allow its dropout layer to be enabled/disabled in the `__call__` method, but only in the `__init__` method, see: https://flax.readthedocs.io/en/latest/_modules/flax/linen/attention.html#MultiHeadDotProductAttention

Let's say we have the following Attention layer:

```python
import flax.linen as nn
from jax import lax, random, numpy as jnp


class DummyAttention(nn.Module):
    num_heads: int

    @nn.compact
    def __call__(self, inputs, deterministic):
        inputs = nn.attention.SelfAttention(
            dropout_rate=0.1,
            num_heads=self.num_heads,
            deterministic=deterministic,
        )(inputs)
        return inputs

model = DummyAttention(num_heads=2)

key1, key2 = random.split(random.PRNGKey(0))
x = jnp.ones((1, 8))
params = model.init({"params": key1, "dropout": key2}, x, True)

# train
print(model.apply(params, x, False, rngs={"dropout": key2}))

# eval
print(model.apply(params, x, True))
```

Now let's say we want to rewrite the module using the `setup` function:

```python
class DummyAttention(nn.Module):
    num_heads: int

   def setup(self):
       self.self_attn = nn.attention.SelfAttention(
            dropout_rate=0.1,
            num_heads=self.num_heads,
            deterministic=deterministic,
        )
    def __call__(self, inputs, deterministic):
        # problem..."deterministic" flag cannot be passed
        inputs = self_attn(inputs)
        return inputs
```

We encounter a problem here, no? We cannot pass `deterministic` dynamically to `SelfAttention` anymore.

Using `setup(self)` with `flax.linen.Dropout` is no problem as the call method accepts a `deterministic` argument.

Would it maybe be possible to add `deterministic` to the `__call__` method of `flax.SelfAttention`? I think it might also better align the API of `flax.Dropout` with `flax.SelfAttention` and allow to dynamically enable/disable dropout when using `setup`.

I'd be happy to open a PR that adds an optional `deterministic` arg to `flax.linen.SelfAttention.__call__`. It could maybe overwrite `SelfAttention's` `deterministic` attribute in case it's passed - what do you think?

**Comment by avital** [comment]
Hi @patrickvonplaten -- this is precisely what has been discussed in #683. (See the proposal doc here: https://github.com/google/flax/blob/ca7d13ccb73d81724de1f106f47730cb2fab96cc/docs/notes/parameters.md). I think that proposal would resolve your concern, if so maybe you could write a quick comment there and we can close this issue. If there's more to do, please let us know.

**Comment by avital** [comment]
Hi @jheek -- assigning this to you, a vote in support of #683

**Comment by patrickvonplaten** [comment]
Hey @avital,

Yes this PR definitely proposes a design that if applied to `SelfAttention` would solve the problem :-) I commented on the PR. Thanks!

**Comment by avital** [comment]
Out of curiosity, in your example, have you considered having `deterministic` also be an attribute in `DummyAttention`? In PyTorch I guess each module has a top-level hidden state -- train vs eval, but you could assume that you module has those defined explicitly at the top and piped through?

**Comment by patrickvonplaten** [comment]
or am I missing something?

In Pytorch one can just switch by doing:

`model.train()` or `model.eval()`

**Comment by avital** [comment]
Yes precisely, PyTorch "solves" this with hard-coded mutable state (train vs eval) which is both incomplete (what if you have a third mode?) and also a bit dangerous (what if you forget to switch between the modes).

@jheek also makes the point that if you have a bound module instance that you tinker with in Colab it's pretty weird to have to clone the instance and replace the variables with the variables from another instance to switch from one mode to another. This can happen for modules defined inside other modules, or for "interactive mutable modules" that will soon land in Flax that allow for easy tinkering in Colab.

So I now see that the change you're asking for here is a good idea. It adds a bit of complexity to the implementation but serves our users better, and one of our guiding philosophies is to "agree" to own more complexity to serve users. A little bit like the "Absorb the complexity" in https://react.christmas/2019/24, but not to get too philosophical here...

**Comment by avital** [comment]
Resolved with #683
